pkgs/build-support/rust/carnix.nix

```diff
@@ -444,167 +460,416 @@ let kernel = buildPlatform.parsed.kernel.name;
 
 in
 rec {
-  aho_corasick_0_6_3 = aho_corasick_0_6_3_ {
+  aho_corasick_0_6_3 = aho_corasick_0_6_3_ rec {
     dependencies = [ memchr_1_0_2 ];
   };
-  ansi_term_0_10_2 = ansi_term_0_10_2_ {};
-  atty_0_2_3 = atty_0_2_3_ {
+  memchr_1_0_2_features."default".from_aho_corasick_0_6_3__default = true;
+  ansi_term_0_10_2 = ansi_term_0_10_2_ rec {};
+  atty_0_2_3 = atty_0_2_3_ rec {
     dependencies = (if kernel == "redox" then [ termion_1_5_1 ] else [])
       ++ (if (kernel == "linux" || kernel == "darwin") then [ libc_0_2_33 ] else [])
       ++ (if kernel == "windows" then [ kernel32_sys_0_2_2 winapi_0_2_8 ] else []);
   };
-  backtrace_0_3_4 = backtrace_0_3_4_ {
+  termion_1_5_1_features."default".from_atty_0_2_3__default = true;
+  libc_0_2_33_features."default".from_atty_0_2_3__default = false;
+  kernel32_sys_0_2_2_features."default".from_atty_0_2_3__default = true;
+  winapi_0_2_8_features."default".from_atty_0_2_3__default = true;
+  backtrace_0_3_4 = backtrace_0_3_4_ rec {
     dependencies = [ cfg_if_0_1_2 rustc_demangle_0_1_5 ]
-      ++ (if (kernel == "linux" || kernel == "darwin") && !(kernel == "fuchsia") && !(kernel == "emscripten") && !(kernel == "darwin") && !(kernel == "ios") then [ backtrace_sys_0_1_16 ] else [])
+      ++ (if (kernel == "linux" || kernel == "darwin") && !(kernel == "fuchsia") && !(kernel == "emscripten") && !(kernel == "darwin") && !(kernel == "ios") then [ backtrace_sys_0_1_16 ]
+      ++ (if lib.lists.any (x: x == "backtrace-sys") features then [backtrace_sys_0_1_16] else []) else [])
       ++ (if (kernel == "linux" || kernel == "darwin") then [ libc_0_2_33 ] else [])
-      ++ (if kernel == "windows" then [ dbghelp_sys_0_2_0 kernel32_sys_0_2_2 winapi_0_2_8 ] else []);
-    features = [ "backtrace-sys" "coresymbolication" "dbghelp" "dbghelp-sys" "dladdr" "kernel32-sys" "libbacktrace" "libunwind" "winapi" ];
-  };
-  backtrace_sys_0_1_16 = backtrace_sys_0_1_16_ {
+      ++ (if kernel == "windows" then [ dbghelp_sys_0_2_0 kernel32_sys_0_2_2 winapi_0_2_8 ]
+      ++ (if lib.lists.any (x: x == "dbghelp-sys") features then [dbghelp_sys_0_2_0] else [])      ++ (if lib.lists.any (x: x == "kernel32-sys") features then [kernel32_sys_0_2_2] else [])      ++ (if lib.lists.any (x: x == "winapi") features then [winapi_0_2_8] else []) else []);
+    features = mkFeatures backtrace_0_3_4_features;
```

Comment:
I have not yet understood the purpose of this function. What does it compute exactly?

Reply:
Previously, Carnix was resolving the dependencies itself. This is not only redundant with what Nix can do, it can also become wrong when some features of some dependencies are enabled only on some platforms.

I'm now using the following trick to resolve features: each package `A` that wants feature f from dependency `B` adds the following line to the "function calls" part of the file:

```
B."f".from_A = true;
```

Then, `mkFeatures` aggregates all such declarations and decides whether "f" should be enabled on package B. Sometimes, feature requests can get more complicated and include `if` statements about the platform, or the activation of another feature.

The "default" feature is trickier, since it is enabled if either no dependent package mentions it, or at least one doesn't want "use-default-features".

Reply:
Is this equivalent to the following code?

```nix
crate_a = buildRustCrate {
   features = ["foo"]; # if "foo" would be a default feature
}; 
crate_b = { 
  dependencies = [ 
  (crate_a.override { features = ["foo" "bar"];} ) # if crate_b has requested "bar" in addition 
  ];
};
```